### ISSUE TITLE
"web-ext run" with custom manifest file

### ISSUE THREAD

**Issue by danburzo** [mozilla/web-ext#2281]
### Is this a feature request or a bug?

This is a feature request (unless it's currently possible and I haven't figured it out). To help with developing / debugging extensions locally, it would be of great help to allow loading a separate `manifest.json` file. In my case, to add `localhost` to the URL patterns for some content scripts, which should not be present in production.

### What is the current behavior?

Currently, I'm adding `<all_urls>` to the content script while working locally, then being careful with each Git commit to not submit the changed manifest file.

### What is the expected or desired behavior?

It would be great to be able to `web-ext run --manifest=manifest.local.json`. Ideally (but not essentially), if it accepted a CommonJS file for the `--manifest` option, I'd be able to extend the main manifest with just the changes I need:

```js
const manifest = require('./manifest.json');
module.exports = {
  ...manifest,
  content_scripts: manifest.content_scripts.map(
    it => ({
      ...it, 
      matches: it.matches.concat('localhost') 
    })
  )
};
```

Alternatively, I'd be glad to know if there's already a solution in the current version of `web-ext`, or otherwise a workflow that doesn't depend on me remembering to not commit `manifest.json` test data.

````
node --version && npm --version && npx web-ext --version
v15.2.0
6.14.10
6.1.0
````

**Comment by rpl** [comment]
@danburzo technically we could, but in that case `web-ext run` can0t just install the extension using the source directory, we would have to build the extension as an xpi (or alternatively create and keep in sync a copy of the source directory) similarly to what we do to support running the extension on a `firefox-android` target.

In the meantime there are a couple of ways to achieve that (and I'm pretty sure that a number of extension developers are doing something similar as part of their own building process), which could be:
- as simple as using a small nodejs script to copy the file into a temporary directory that you then point `web-ext run` to (and maybe along with watching the source directory and keep in sync the temporary directory)
  - `web-ext run` does support a `--watch-file` option that can be used to make web-ext to reload the extension only when that particular file is updated, which makes it easier to coordinate custom building steps in the extension with the `web-ext run` autoreload on changes feature.
- or using some more complex building and bundling tools (which is a more than reasonable solution if the extension does already need a bundling or bundling step also for other reasons)  
  - e.g. FirefoxColor does exactly what you describe [using a webpack plugin](https://github.com/mozilla/FirefoxColor/blob/master/webpack.extension.js#L34-L102)

**Comment by danburzo** [comment]
Makes sense! Thanks for the pointers, I will try one of the options you propose.

**Comment by fregante** [comment]
This looks like something the build tool should do, and in my build that's where the same manifest.json configuration already happens. Just my opinion, I don't mind either way.

**Comment by rpl** [comment]
> This looks like something the build tool should do, and in my build that's where the same manifest.json configuration already happens. Just my opinion, I don't mind either way.

I don't disagree, as can be guessed by the pointers I included in my previous comment, but I haven't closed it as wontfix because I want to discuss about it with my teammates in the next bug triaging meeting (especially because of the "small bits shared" between this issue and #2277)

**Comment by rpl** [comment]
We discussed this in our previous triage meeting and agreed to wontfix this (the rationale doesn't differ from the one described in @fregante comment https://github.com/mozilla/web-ext/issues/2281#issuecomment-875780695, this is better solved as part of the extension build process).

**Comment by davidmaxwaterman** [comment]
...I came across this issue because I want a different manifest when I am testing. I had it in mind to keep the src/manifest.json as the one I use for building the extension for production, but also have a test/manifest.json file that also includes the extra files needed for testing. I put these extra files also in the src/ tree, and trust that they are not included in the production build because they're not mentioned in src/manifest.json.
To test my extension, I would have liked to have simply done `--override-manifest=test/manifest.json --source-dir=src/` and web-ext would ignore src/manifest.json and instead use test/manifest.json.
From reading this issue, it seems like I need to add a step to my `test` command in package.json to copy test/manifest.json to src/manifest.json before running the tests, and then `reset` src/manifest.json again afterwards.
WDYT?

**Comment by Rob--W** [comment]
I also have extensions with build/browser-specific manifest.json files. What I usually do is to compile to browser-specific output directory, e.g. `dist/firefox` and `dist/chrome`, and then do `web-ext` run from there. "Building" can also be as simple as just copying the files and copying a specific manifest.json file.

It's not clear from your description whether you'd be modifying the source files directly. While that could work, it is not ideal because you'd be touching source files, and if there is no version control, the originals can be lost. So I still recommend the build approach that I sketched before.

For your specific use case, a fix for #1107 would be preferable. Patches are welcome ;)

**Comment by davidmaxwaterman** [comment]
Yeah. There is the case where you don't actually have a build step, and just run from source, so that would be a plain 'cp' build, which I find a little pointless and prone to errors.

I wouldn't modify any source files - I do have some envs that vite replaces during build, and/or are replaced by the development server I use (@web/dev-server)...or something like that anyway.

It seems that [vite](https://github.com/samrum/create-vite-plugin-web-extension/blob/main/template/base/package.json#L10) doesn't work on FF, so I ended up building anyway, and using wds to serve the test scripts/pages.

IINM, someone already did a fix for #1107 on a [fork](https://github.com/piroor/web-ext/wiki/Installation), so it would be a little pointless to do the work again (although I noticed that he/she did it differently to how I would have...ie he/she did it using a single `--source-dir` option with multiple dirs delimited by commas, while I would have specified multiple `--source-dir` options each with a single dir (so that there isn't a problem with directory names containing commas, which is, admittedly, unlikely)).

Anyway, it just felt like I was stymie by one limitation, and forced into another approach, only to be stymied again by another limitation. Such is life :)

The project in question is a prototype and has been shelved for now, so it's a little moot from my point of view. I ended up just copying a test manifest (as I described) and letting rollup sort out what files are needed in the build, presumably using tree-shaking, which is kind of cool. I did actually take a look at the build output and it indeed did exclude the test files when I built with the non-test manifest.json, so that's kind of nice :)
